pennylane/transforms/control.py

```diff
@@ -0,0 +1,191 @@
+                )
+    return new_tape
+
+
+class ControlledOperation(Operation):
+    """A Controlled Operation.
+
+    Unless you are a Pennylane plugin developer, **you should NOT directly use this class**,
+    instead, use the ``qml.ctrl`` method.
+
+    The ``ControlledOperation`` is a container class that defines a set of operations that
+    should by applied relative to a single control wire or a list of control wires.
+
+    Certain simulators and quantum computers can take advantage of the controlled gate sparsity,
+    while other devices must rely on the op-by-op decomposition defined by the ``op.expand``
+    method.
+
+    Attributes:
+        subtape (QuantumTape): The tape that defines the underlying operation.
+        control_wires (Wires): The control wires.
+
+    Args:
+        tape: A QuantumTape. This tape defines the unitary that should be applied relative
+            to the control wires.
+        control_wires: A wire or set of wires.
+    """
+
+    par_domain = "A"
+    num_wires = AnyWires
+    num_params = property(lambda self: self.subtape.num_params)
+
+    def __init__(self, tape, control_wires, do_queue=True):
+        self.subtape = tape
+        self.control_wires = Wires(control_wires)
+        wires = self.control_wires + tape.wires
+        super().__init__(*tape.get_parameters(), wires=wires, do_queue=do_queue)
+
+    def expand(self):
+        tape = self.subtape
+        for wire in self.control_wires:
+            tape = expand_with_control(tape, wire)
+        # TODO(chase): Do we need to re-queue these ops?
+        return tape
+
+    def adjoint(self, do_queue=False):
+        def requeue_tape(tape):
+            for op in tape.operations:
+                op.queue()
+
+        with QuantumTape(do_queue=False) as new_tape:
+            # Execute all ops adjointed.
+            adjoint(requeue_tape)(self.subtape)
+        return ControlledOperation(new_tape, self.control_wires, do_queue=do_queue)
+
+
+register_control(
+    ControlledOperation,
+    lambda op, wires: ControlledOperation(
+        tape=op.subtape,
+        control_wires=Wires(wires) + op.control_wires,
+    ),
+)
+
+
+def ctrl(fn, control):
+    """Create a method that applies a controlled version of the provided method.
+
+    Args:
+        fn (function): Any python function that applies pennylane operations.
+        control (Wires): The control wire(s).
+
+    Returns:
+        A new function that applies the controlled equivalent of `fn`.
+
+    **Example**
+
+    .. code-block:: python3
+
+        def ops(params):
+            qml.RX(params[0], wires=0)
+            qml.RZ(params[1] wires=3)
+
+        ops1 = qml.ctrl(ops, control=1)
+        ops2 = qml.ctrl(ops, control=2)
+
+        @qml.qnode(dev)
+        def my_circuit():
+            ops1(params=[0.123, 0.456])
+            ops1(params=[0.789, 1.234])
+            ops2(params=[2.987, 3.654])
+            ops2(params=[2.321, 1.111])
+            return qml.state()

```

Comment:
Docstring looks great! https://pennylane--1157.org.readthedocs.build/en/1157/code/api/pennylane.ctrl.html

I had two questions while reading:

- Instead of re-writing the code-block below in expanded form, could we just `qml.draw()` the expanded circuit?
- And as soon as I had that thought, I realised that `qml.draw()` will likely be broken on QNodes that include controlled ops?

Reply:
Yeah I didn't add it because `draw` doesn't work yet. Supporting draw seemed like a whole ordeal so I wanted to just get this minimal PR working first.

Reply:
Yep, agree, it is likely it's own PR. What I might suggest though is quickly adding validation to `qml.draw()`, just so it spits out a reasonable error message if the circuit has `ControlledOps` in it.

I'm not 100% sure if we'll have a chance to update `qml.draw()` before the next release, and I worry about catching user questions.

Reply:
Yeah... there is a 100% chance someone will try it before we fix it.

Reply:
Aw, I just tested it and right now it will expand the operation before drawing it, so drawing does technically work.

I think that should be fine for now, especially since no devices take advantage of this yet and they always execute expanded anyway

Reply:
wait, that's awesome!